// @radix-ui/react-toggle/dist/index.mjs.js
'use client';
import * as React from 'react';
import { composeEventHandlers } from '../../primitive/dist/index.mjs.js';
import { useControllableState } from '../../react-use-controllable-state/dist/index.mjs.js';
import { Primitive } from '../../react-primitive/dist/index.mjs.js';
import { jsx } from 'react/jsx-runtime';

var NAME = "Toggle";
var Toggle = React.forwardRef((props, forwardedRef) => {
  const { pressed: pressedProp, defaultPressed = false, onPressedChange, ...buttonProps } = props;
  const [pressed = false, setPressed] = useControllableState({
    prop: pressedProp,
    onChange: onPressedChange,
    defaultProp: defaultPressed
  });
  return /* @__PURE__ */ jsx(
    Primitive.button,
    {
      type: "button",
      "aria-pressed": pressed,
      "data-state": pressed ? "on" : "off",
      "data-disabled": props.disabled ? "" : void 0,
      ...buttonProps,
      ref: forwardedRef,
      onClick: composeEventHandlers(props.onClick, () => {
        if (!props.disabled) {
          setPressed(!pressed);
        }
      })
    }
  );
});
Toggle.displayName = NAME;

export { Toggle };
